### ISSUE TITLE
create "v2" legacy mode configuration option

### ISSUE THREAD

**Issue by job** [irrdnet/irrd4#137]
It appears there are some discrepancies between how `!i` works in irrd v2 and irrd v3

I'd like to propose that as a transitional mechanism a configuration flag is added to the daemon to allow it to operate in either `v2` or `v3` mode:

in v2 mode:

`!i` returns only ipv4 prefixes when applied to `route-set` objects
`!i6` returns only ipv6 prefixes when applied to `route-set` objects

in v3 mode (the default)

`!i` returns both ipv4 and ipv6 prefixes when applied to `route-set` objects
`!i6` does not exist

**Comment by job** [comment]
As discussed:

- add `!i6` as command to v4, so that query command exists in both v2, v3, and v4 mode
- make a configuration flag that influences whether `!i` returns v6 prefixes or not for `route-sets`

**Comment by mxsasha** [comment]
@job I wanted to verify this feature, but it seems `!i6` is not functioning at all on rr.ntt.net? This raises questions, which are a release blocker.

For this object, both `!i` and `!i6` only return from `members`:
```
>>> !sAFRINIC
C
>>> RS-HomeFind24
route-set:      RS-HomeFind24
descr:          HomeFind24 Route Set
members:        80.88.11.0/24, 102.165.188.0/24
mp-members:     80.88.11.0/24, 102.165.188.0/24, 2001:43f8:1070::/48
tech-c:         PTO2-AFRINIC
admin-c:        PTO2-AFRINIC
mnt-by:         HomeFind24-MNT
changed:        ***@property24.com 20171205
source:         AFRINIC

>>> !iRS-HomeFind24
A31
102.165.188.0/24 80.88.11.0/24
C
>>> !i6RS-HomeFind24
A31
102.165.188.0/24 80.88.11.0/24
C
```

Same here:
```
>>> RS-BBS
route-set:      RS-BBS
descr:          BURUNDI BACKBONE SYSTEM CUSTOMERS AND OWN PREFIXES
members:        41.79.224.0/22
members:        154.73.40.0/22
members:        102.134.96.0/20
members:        RS-BBS-CUSTOMERS
mp-members:     2c0f:f6a8::/32
tech-c:         ACN1-AFRINIC
admin-c:        ACN1-AFRINIC
mnt-by:         BBS-MNT
changed:        ***@bbs.bi 20181207
changed:        ***@bbs.bi 20181210
source:         AFRINIC

>>> !iRS-BBS
A63
102.134.96.0/20 154.73.40.0/22 41.79.224.0/22 RS-BBS-CUSTOMERS
C
>>> !i6RS-BBS
A63
102.134.96.0/20 154.73.40.0/22 41.79.224.0/22 RS-BBS-CUSTOMERS
C
```

Or this one:
```
> !i6AS12654:RS-RIS,1
A649
185.0.0.0/16 185.1.0.0/21 23.128.24.0/24 23.128.25.0/25 23.128.25.240/28 84.205.64.0/24 84.205.65.0/24 84.205.66.0/24 84.205.67.0/24 84.205.68.0/24 84.205.69.0/24 84.205.70.0/24 84.205.71.0/24 84.205.72.0/24 84.205.73.0/24 84.205.74.0/24 84.205.75.0/24 84.205.76.0/24 84.205.77.0/24 84.205.78.0/24 84.205.79.0/24 84.205.80.0/24 84.205.81.0/24 84.205.82.0/24 84.205.83.0/24 84.205.84.0/24 84.205.85.0/24 84.205.86.0/24 84.205.87.0/24 84.205.88.0/24 84.205.89.0/24 84.205.90.0/24 84.205.91.0/24 84.205.92.0/24 84.205.93.0/24 84.205.94.0/24 84.205.95.0/24 93.175.146.0/24 93.175.147.0/24 93.175.148.0/24 93.175.149.0/24 93.175.150.0/24 93.175.151.0/24
```

In other words, the `!i6` query is accepted, but does not actually seem to work in IRRd as run by NTT. Do we still need this feature? I will wait for you to confirm what the behaviour should be, because if we continue as planned, we will actually create a potential new incompatibility with rr.ntt.net.

**Comment by troy2914** [comment]
!i6 was only for as-sets, not route-sets.

**Comment by mxsasha** [comment]
Are you sure you're not thinking of `!6`? Because `!i` for as-sets will never return IPv4 or IPv6 prefixes, so the distinction there would have no effect.

**Comment by troy2914** [comment]
well the intent was !i + route_set only returns v4 records.
And !i6 + route_set only returns v6 records.

As you note though v2 does not respond with v6 records to !i6.
v3 does not respond to !i6 at all.

I am reconsidering whether this ticket is needed.   For our use there are only 2 customers currently using route-set's.  and for both of those the v6 is listed as an as-set (probably because of the above noted response to !i6 in v2).   However both of those route-sets include mp-members with v6 records.

It seems like need a option to disable returning mp-members if running in v2 mode.  I really don't care about a !i6 command.

**Comment by troy2914** [comment]
Correction,  in v2 mode !i of route-set's should only return ipv4 addresses.   There may be ipv6 in the members that we want to filter, and ipv4 in mp-members that we want to include.

We do not need a v3 mode, and we do not need an `!i6` command

**Comment by mxsasha** [comment]
> Correction, in v2 mode !i of route-set's should only return ipv4 addresses. There may be ipv6 in the members that we want to filter, and ipv4 in mp-members that we want to include.

This makes sense, however, it is not how rr.ntt.net currently functions:
```
>>> RS-MP-MEMBERS-TEST
route-set:      RS-MP-MEMBERS-TEST
members:        1.1.1.0/24
mp-members:     2.2.2.0/24, 2001:db8::/32
tech-c:         DUMY-RIPE
admin-c:        DUMY-RIPE
mnt-by:         sr42-mnT
created:        2019-01-08T09:06:28Z
last-modified:  2019-01-08T09:06:28Z
source:         RIPE
remarks:        ****************************
remarks:        * THIS OBJECT IS MODIFIED
remarks:        * Please note that all data that is generally regarded as personal
remarks:        * data has been removed from this object.
remarks:        * To view the original object, please query the RIPE Database at:
remarks:        * http://www.ripe.net/whois
remarks:        ****************************

>>> !iRS-MP-MEMBERS-TEST
A11
1.1.1.0/24
C
>>> !i6RS-MP-MEMBERS-TEST
A11
1.1.1.0/24
C
```

It only looks at members, and ignores all mp-members. But, it gets weirder:

```
>>> !iRS-CZ-NIC,1
A299
185.43.132.0/22 185.43.132.0/24 185.43.133.0/24 185.43.134.0/24 192.88.99.0/24 193.0.14.0/23 193.0.14.0/24 193.17.47.0/24 193.29.206.0/24 194.0.12.0/24 194.0.13.0/24 194.0.14.0/24 199.253.182.0/24 199.253.183.0/24 199.6.4.0/24 199.7.82.0/23 199.7.83.0/24 200.1.121.0/24 200.7.4.0/24 217.31.192.0/20
C

>>> !i6RS-CZ-NIC,1
A380
185.43.132.0/22 185.43.132.0/24 185.43.133.0/24 185.43.134.0/24 192.88.99.0/24 193.0.14.0/23 193.0.14.0/24 193.17.47.0/24 193.29.206.0/24 194.0.12.0/24 194.0.13.0/24 194.0.14.0/24 199.253.182.0/24 199.253.183.0/24 199.6.4.0/24 199.7.82.0/23 199.7.83.0/24 200.1.121.0/24 200.7.4.0/24 2001:1488::/32 2001:678:10::/48 2001:678:11::/48 2001:678:1::/48 2001:678:f::/48 217.31.192.0/20
C
```

The IPv6 routes returned here are route6 objects with a member-of set to RS-CZ-NIC. So it appears that in v2:
* `!i` includes member attributes and route objects with an appropriate member-of/mbrs-by-ref
* `!i6` includes member attributes and route+route6 objects with an appropriate member-of/mbrs-by-ref
* Neither query considers mp-members, whether that contains IPv4 or IPv6 prefixes.

In other words, different, equally valid, ways of recording the exact same information in RPSL, will result in different query outputs in v2. This seems so broken to me that we should not seek to exactly replicate existing behaviour, but go with your suggestion that:
* v2 mode setting will cause `!i` to only return IPv4 prefixes, whether they are in members, mp-members, or a route with member-of/mbrs-by-ref.
* Not implement `!i6`

(IPv6 addresses in members attributes are never a concern, as they are not permitted and no databases seem to allow it, see #53).

**Comment by job** [comment]
Sounds good to me
